Show related hpa in dc description

kargakis opened the pull request:
Changes in this PR:
* Addition of namespace
* Autoscaling info
* Factored out custom interfaces

```
[vagrant@localhost sample-app]$ oc describe dc
Name:		docker-registry
Namespace:	default
Created:	6 minutes ago
Labels:		docker-registry=default
Annotations:	<none>
Latest Version:	2
Selector:	docker-registry=default
Replicas:	1
Autoscaled:	min: 1, max: 3, targeting 80% cpu
Triggers:	Config
Strategy:	Rolling
Template:
  Labels:	docker-registry=default
  Containers:
  registry:
    Image:	openshift/origin-docker-registry:latest
    Port:	5000/TCP
    QoS Tier:
      memory:	BestEffort
      cpu:	BestEffort
    Liveness:	http-get http://:5000/healthz delay=10s timeout=5s period=10s #success=1 #failure=3
[ ... ]
```

@smarterclayton @ironcladlou @DirectXMan12 PTAL

cc: @jwforres @fabianofranz

jwforres commented:
cc @spadgett

On Mon, Mar 14, 2016 at 12:32 PM, Michail Kargakis <notifications@github.com
> wrote:

> $ oc describe dc/docker-registry# [ ... ]
> Autoscaled: min: 1, max: 5, targeting 80% cpu# [ ... ]
>
> Also I couldnt stand the temptation to refactor the describer interfaces:)
>
> @smarterclayton <https://github.com/smarterclayton> @ironcladlou
> <https://github.com/ironcladlou> @DirectXMan12
> <https://github.com/DirectXMan12> PTAL
>
> cc: @jwforres <https://github.com/jwforres> @fabianofranz
> <https://github.com/fabianofranz>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/openshift/origin/pull/7995
> Commit Summary
>
>    - oc: remove special-cased dc describer interfaces
>    - oc: print autoscaling info for deploymentconfigs
>
> File Changes
>
>    - *M* pkg/cmd/cli/describe/deployments.go
>    <https://github.com/openshift/origin/pull/7995/files#diff-0> (241)
>    - *M* pkg/cmd/cli/describe/describer_test.go
>    <https://github.com/openshift/origin/pull/7995/files#diff-1> (41)
>
> Patch Links:
>
>    - https://github.com/openshift/origin/pull/7995.patch
>    - https://github.com/openshift/origin/pull/7995.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7995>.
>

kargakis commented:
[test]

deads2k commented on pkg/cmd/cli/describe/deployments.go:
```diff
@@ -28,99 +29,25 @@ import (
-	events, err := d.client.listEvents(deploymentConfig)
+	events, err := d.kubeClient.Events(deploymentConfig.Namespace).Search(deploymentConfig)
 	if err != nil {
 		return "", err
```
This should not have ever failed on this.  Fix it while you're here.  An error on secondary retrieval should indicate that the lookup failed, but not crater the describe

deads2k commented on pkg/cmd/cli/describe/deployments.go:
```diff
@@ -224,6 +207,34 @@ func printHook(prefix string, hook *deployapi.LifecycleHook, w io.Writer) {
 	}
 }
 
+func printAutoscalingInfo(config *deployapi.DeploymentConfig, kclient kclient.Interface, w *tabwriter.Writer) {
```
Seems like you want a `GroupResource` and a `name`, right?  That would make this re-usable instead of having to create separate ones for each type.

kargakis replied:
Right. Eventually this helper should move upstream.

deads2k commented:
Where does the current scale output in the describer?  Nearby?

kargakis commented:
> Where does the current scale output in the describer? Nearby?

Hm, nice shot in the dark. We don't print the desired size for the dc. I can add it. We print each deployment separately though.

spadgett commented:
Should we specifically say it's a percent of CPU request?

deads2k commented:
> Hm, nice shot in the dark. We don't print the desired size for the dc. I can add it. We print each deployment separately though.

Yeah, knowing what its autoscaling to is a lot more useful if I can also find out what my current scale is.

smarterclayton commented:
Not for 1.2

kargakis commented:
Actually we do print the replicas, nvmd. For some reason it's buried under "Template"

deads2k commented:
> Actually we do print the replicas, nvmd. For some reason it's buried under "Template"

Can you update the description with the output?  I think the lines should be close together.

kargakis commented:
Changes so far in this PR:
* Addition of namespace
* Moved Latest Version next to the deployment history
* Removed the "Template" wording which was inaccurate and made "Replicas" and "Selector" first class info. Moved the Selector next to the object meta.
* Autoscaling info (I still don't like the output here, any suggestions?)
```
$ oc describe dc/frontend
Name:		frontend
Namespace:	test
Created:	About an hour ago
Labels:		template=application-template-stibuild
Annotations:	<none>
Selector:	name=frontend
Replicas:	2
Autoscaled:	min: 1, max: 4, targeting 80% cpu
Triggers:	Image(origin-ruby-sample@latest, auto=true), Config
Strategy:	Rolling
		  Pre-deployment hook (pod type, failure policy: Abort):
		    Container:	ruby-helloworld
		    Command:	/bin/true
		    Env:	CUSTOM_VAR1=custom_value1
		  Post-deployment hook (pod type, failure policy: Ignore):
		    Container:	ruby-helloworld
		    Command:	/bin/false
		    Env:	CUSTOM_VAR2=custom_value2
Containers:
  ruby-helloworld:
    Image:	172.30.40.231:5000/test/origin-ruby-sample@sha256:d22fc33c3e5d2fa7fc28c9c0abd07d0b5bc0f8144b35fab0f0a19cb9fecd997c
    QoS Tier:
      cpu:	BestEffort
      memory:	BestEffort
    Environment Variables:
      ADMIN_USERNAME:	adminGK3
      ADMIN_PASSWORD:	6rRQYNFg
      MYSQL_USER:	userSXV
      MYSQL_PASSWORD:	iJSpkjo8
      MYSQL_DATABASE:	root
Latest Version:		2
Deployment #2 (latest):
	Name:		frontend-2
	Created:	2 minutes ago
	Status:		Complete
	Replicas:	2 current / 2 desired
	Selector:	deployment=frontend-2,deploymentconfig=frontend,name=frontend
	Labels:		openshift.io/deployment-config.name=frontend,template=application-template-stibuild
	Pods Status:	2 Running / 0 Waiting / 0 Succeeded / 0 Failed
Deployment #1:
	Created:	about an hour ago
	Status:		Complete
	Replicas:	0 current / 0 desired
Events:
  FirstSeen	LastSeen	Count	From				SubobjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  1h		1h		1	{deploymentconfig-controller }			Normal		DeploymentCreated	Created new deployment "frontend-1" for version 1
  1h		1h		1	{deployer }					Warning		FailedUpdate		Error updating deployment test/frontend-1 status to Pending
  2m		2m		1	{deploymentconfig-controller }			Normal		DeploymentCreated	Created new deployment "frontend-2" for version 2
  2m		2m		1	{deployer }					Warning		FailedUpdate		Error updating deployment test/frontend-2 status to Pending
```

kargakis commented:
[test]

kargakis commented:
[test]

fabianofranz commented:
Few ideas:

```
Autoscaling: between 1 and 4 replicas, triggered at 80% CPU usage

Autoscaling: from 1 to 4 replicas (currently 2), triggered at 80% CPU usage

Autoscaling: 2 replicas running (1 to 4 allowed), triggered at 80% CPU usage

Autoscaling: 1~4 (currently 2) replicas, triggered at 80% CPU usage
```

smarterclayton commented:
latest state?

kargakis commented:
Will see what's needed on Monday

On Sat, May 14, 2016 at 4:33 PM, Clayton Coleman <notifications@github.com>
wrote:

> latest state?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7995#issuecomment-219223466>
>

kargakis commented:
@fabianofranz @ironcladlou @smarterclayton this is ready for another look

smarterclayton commented on pkg/cmd/cli/describe/helpers.go:
```diff
@@ -129,6 +129,7 @@ func FormatRelativeTime(t time.Time) string {
 
 func formatMeta(out *tabwriter.Writer, m api.ObjectMeta) {
 	formatString(out, "Name", m.Name)
+	formatString(out, "Namespace", m.Namespace)
```
I don't want to print namespace unless we really have to - everything is effectively in a namespace.  Reason?

kargakis replied:
> Reason?

All kube describers already do this.

smarterclayton commented:
OK. Can you add a test case for the deployment describer?  I realized it's not covered at all.

kargakis commented:
> OK. Can you add a test case for the deployment describer? I realized it's not covered at all.

We actually have a test for it but it is buried with the rest of the describer tests. Moved it to a separate file since we have the dc describer separate too. Added testing for the hpa output. PTAL

smarterclayton commented:
LGTM [merge] thanks
